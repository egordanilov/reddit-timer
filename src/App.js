import React from 'react';
import { Routes, Route } from 'react-router-dom';
import Home from './Home';
import Search from './Search';
import GlobalStyles from './styles/GlobalStyles';

function App() {
  return (
    <>
      <GlobalStyles />
      <header>Header</header>
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="/search" element={<Search />} />
      </Routes>
      <footer>Footer</footer>
    </>
  );
}

export default App;
