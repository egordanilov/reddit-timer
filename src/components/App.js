import React from 'react';
import { Outlet, Routes, Route } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import GlobalStyles from '../styles/GlobalStyles';
import Header from '../UI/Header';
import Footer from '../UI/Footer';
import Home from '../pages/Home';
import Search from '../pages/Search';
import Terms from '../pages/Terms';
import theme from '../styles/theme';

function App() {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      <Header />
      <Routes>
        <Route path="/">
          <Route path="search" element={<Search />}>
            <Route path=":subreddit" element={<Search />} />
            <Route index element={<Search />} />
          </Route>
          <Route path="/terms" element={<Terms />} />
          <Route index element={<Home />} />
        </Route>
      </Routes>
      <Outlet />
      <Footer />
    </ThemeProvider>
  );
}

export default App;
