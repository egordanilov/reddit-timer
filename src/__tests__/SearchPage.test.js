import React from 'react';
import {
  render, screen, within, waitForElementToBeRemoved, waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import App from '../components/App';
import { LocationDisplay } from './App.test';

const setup = (initialPath) => {
  // access history as described in the docs
  // https://reactrouter.com/web/guides/testing/checking-location-in-tests
  let history;
  const view = render(
    <MemoryRouter initialEntries={[initialPath]}>
      <App />
      <Routes>
        <Route
          path="*"
          render={(props) => {
            history = props.history;
            return null;
          }}
        />
      </Routes>
      <LocationDisplay />
    </MemoryRouter>,
  );
  return { ...view, history };
};

test('initializes the input value from the URL and updates the URL with the new subreddit on submit', async () => {
  setup('/search/angular');
  const headerLink = screen.getByRole('link', { name: 'Search' });
  const submitButton = screen.getByRole('button', { name: 'SEARCH' });
  const locationDisplay = screen.getByTestId('location-display');
  const subredditInput = screen.getByLabelText('r /');
  // Expect the form input value to initialize from URL
  expect(subredditInput.value).toBe('angular');
  // Clear the form input, type in reactjs, submit the form, expect the path to be reactjs
  await userEvent.clear(subredditInput);
  await userEvent.type(subredditInput, 'reactjs');
  await userEvent.click(submitButton);
  expect(locationDisplay).toHaveTextContent(subredditInput.value);
  // Click on Search link in the header and expect a default input param and url to be javascript
  await userEvent.click(headerLink);
  expect(subredditInput.value).toBe('javascript');
  expect(locationDisplay).toHaveTextContent('/search/javascript');
});

describe('heatmap', () => {
  test('loads and reddits top posts for a subreddit in url', async () => {
    setup('/search/angular');
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.getByTestId('loading-spinner'));
    const heatmap = screen.getByTestId('heatmap');
    const cells = within(heatmap).getAllByRole('button');
    expect(cells.length).toEqual(7 * 24);
    const cellsContent = cells.map((cell) => cell.innerHTML);
    expect(cellsContent).toMatchSnapshot();
    expect(screen.getByTestId('timezone')).toHaveTextContent('All times are shown in your timezone: America/Chicago');
  });

  test('cell highlights on click', async () => {
    setup('/search/reactjs');
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.getByTestId('loading-spinner'));
    const heatmap = screen.getByTestId('heatmap');
    const cells = within(heatmap).getAllByRole('button');
    // day and hour selected by default inside Heatmap [activeCell, setActiveCell]
    expect(cells[12]).toHaveStyle('border: 2px solid #93918F');
    const cellToClick = cells[15];
    expect(cellToClick).toHaveStyle('border: 0px solid #93918F');
    await userEvent.click(cellToClick);
    expect(cellToClick).toHaveStyle('border: 2px solid #93918F');
  });
  test('renders error message', async () => {
    setup('/search/failing-request');
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.getByTestId('loading-spinner'));
    expect(screen.getByText('Failed to fetch, check internet connection and subreddit name')).toBeInTheDocument();
  });
});

async function clickOnHeatMapCellWithValue(value) {
  await waitFor(() => screen.getByTestId('heatmap'));
  const heatmap = await screen.getByTestId('heatmap');
  const cellWithValue = within(heatmap).getAllByText(value)[0];
  userEvent.click(cellWithValue);
}

describe('posts table', () => {
  test('the posts table is shown when a box in the heatmap has been clicked', async () => {
    setup('/search/javascript');
    await clickOnHeatMapCellWithValue('1');
    expect(screen.getAllByRole('columnheader').length).toEqual(5);
  });
  test('the posts table is not shown when there are no posts for the selected weekday/hour', async () => {
    setup('/search/javascript');
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.getByTestId('loading-spinner'));
    await clickOnHeatMapCellWithValue('0');
    expect(screen.queryByRole('table')).not.toBeInTheDocument();
  });
  test("the table shows the post's title, time, score (number of upvotes), number of comments and author and sorted by time when post was created", async () => {
    setup('/search/javascript');
    await clickOnHeatMapCellWithValue('4');

    const table = screen.getByRole('table');
    /* slice the table header row */
    const tableRows = within(table).getAllByRole('row').slice(1);
    const tableContent = tableRows.map((row) => {
      const cells = within(row).getAllByRole('cell');
      const titleLink = within(cells[0]).getByRole('link');
      const authorLink = within(cells[4]).getByRole('link');
      return {
        title: titleLink.innerHTML,
        href: titleLink.href,
        time: cells[1].innerHTML,
        score: cells[2].innerHTML,
        numComments: cells[3].innerHTML,
        author: authorLink.innerHTML,
        authorHref: authorLink.href,
      };
    });

    expect(tableContent).toMatchSnapshot();
  });

  test('shows no link for a deleted user', async () => {
    setup('/search/javascript');
    await waitFor(() => screen.getByTestId('heatmap'));
    const heatmap = await screen.getByTestId('heatmap');
    const cells = within(heatmap).getAllByRole('button');
    await userEvent.click(cells[77]);
    const table = await screen.getByRole('table');
    const columnWithDeletedUser = within(table).getByRole('cell', { name: /\[deleted\]/i });
    expect(within(columnWithDeletedUser).queryByRole('link')).not.toBeInTheDocument();
    expect(columnWithDeletedUser.innerHTML).toBe('[deleted]');

    screen.debug();
  });
});
