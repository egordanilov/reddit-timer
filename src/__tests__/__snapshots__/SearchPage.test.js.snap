// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`heatmap loads and reddits top posts for a subreddit in url 1`] = `
Array [
  "1",
  "2",
  "1",
  "4",
  "2",
  "4",
  "3",
  "4",
  "4",
  "7",
  "7",
  "2",
  "2",
  "2",
  "6",
  "0",
  "0",
  "1",
  "1",
  "0",
  "3",
  "1",
  "0",
  "1",
  "1",
  "2",
  "4",
  "4",
  "4",
  "4",
  "4",
  "6",
  "3",
  "4",
  "6",
  "4",
  "6",
  "6",
  "1",
  "5",
  "3",
  "3",
  "2",
  "2",
  "3",
  "0",
  "1",
  "1",
  "1",
  "5",
  "2",
  "1",
  "1",
  "1",
  "2",
  "3",
  "11",
  "5",
  "5",
  "9",
  "5",
  "2",
  "9",
  "3",
  "4",
  "2",
  "1",
  "3",
  "0",
  "1",
  "1",
  "0",
  "1",
  "1",
  "3",
  "2",
  "3",
  "4",
  "3",
  "0",
  "6",
  "9",
  "5",
  "2",
  "6",
  "6",
  "2",
  "5",
  "1",
  "3",
  "3",
  "2",
  "1",
  "2",
  "0",
  "2",
  "3",
  "1",
  "1",
  "4",
  "3",
  "6",
  "2",
  "6",
  "1",
  "5",
  "7",
  "7",
  "9",
  "4",
  "0",
  "5",
  "2",
  "2",
  "3",
  "3",
  "0",
  "0",
  "0",
  "0",
  "2",
  "1",
  "1",
  "3",
  "2",
  "3",
  "2",
  "6",
  "5",
  "6",
  "12",
  "7",
  "2",
  "2",
  "3",
  "3",
  "1",
  "5",
  "6",
  "3",
  "0",
  "3",
  "2",
  "1",
  "0",
  "2",
  "2",
  "3",
  "0",
  "3",
  "3",
  "4",
  "5",
  "4",
  "3",
  "9",
  "4",
  "1",
  "1",
  "3",
  "2",
  "1",
  "1",
  "1",
  "1",
  "1",
  "3",
  "2",
]
`;

exports[`posts table the table shows the post's title, time, score (number of upvotes), number of comments and author and sorted by time when post was created 1`] = `
Array [
  Object {
    "author": "ConfidentMushroom",
    "authorHref": "https://reddit.com/u/ConfidentMushroom",
    "href": "https://reddit.com/r/javascript/comments/s5826l/new_chrome_devtools_feature_record_replay_and/",
    "numComments": "23",
    "score": "304",
    "time": "3:16 am",
    "title": "New Chrome DevTools feature: record, replay, and measure user interactions",
  },
  Object {
    "author": "shane_il",
    "authorHref": "https://reddit.com/u/shane_il",
    "href": "https://reddit.com/r/javascript/comments/qeny0z/askjs_looking_for_an_os_agnostic_search_module/",
    "numComments": "11",
    "score": "302",
    "time": "3:18 am",
    "title": "[AskJS] Looking for an OS agnostic search module for node.",
  },
  Object {
    "author": "jon_abides",
    "authorHref": "https://reddit.com/u/jon_abides",
    "href": "https://reddit.com/r/javascript/comments/qjl9lb/ive_created_a_vanilla_js_lib_that_generates/",
    "numComments": "34",
    "score": "216",
    "time": "3:29 am",
    "title": "Iâ€™ve created a Vanilla JS lib that generates animated waves",
  },
  Object {
    "author": "NiceGuy-n2",
    "authorHref": "https://reddit.com/u/NiceGuy-n2",
    "href": "https://reddit.com/r/javascript/comments/wib2ra/express_like_api_for_bunjs_just_for_fun/",
    "numComments": "7",
    "score": "75",
    "time": "3:47 am",
    "title": "Express like api for bunjs ( Just for fun)",
  },
]
`;
